File: syntavra_runtime/migration_plan_read_only_contract.py
from __future__ import annotations

import hashlib
import json
import math
import os
import re
import sqlite3
import stat
from pathlib import Path
from typing import Any, Final

CONTRACT_VERSION: Final = 1
SCHEMA_VERSION: Final = 1
CAPABILITY: Final = "migration.plan"
ROUTE: Final = "migration.plan"
MAXIMUM_DATABASE_BYTES: Final = 64 * 1024 * 1024
MAXIMUM_PATH_BYTES: Final = 4096
INPUT_PROFILE: Final = "project-bound-quiescent-migration-sqlite-v1"
INPUT_FORMAT: Final = "canonical-project-relative-path"

_SIDECAR_SUFFIXES: Final = ("-journal", "-shm", "-wal")
_IDENTITY_RE: Final = re.compile(r"^[0-9a-f]{64}$")
_EXPECTED_COLUMNS: Final = (
    ("version", "INTEGER", 0, 1),
    ("name", "TEXT", 1, 0),
    ("identity", "TEXT", 1, 0),
    ("applied_at", "REAL", 1, 0),
)


class MigrationPlanReadOnlyError(RuntimeError):
    pass


def _error(code: str, cause: BaseException | None = None) -> MigrationPlanReadOnlyError:
    error = MigrationPlanReadOnlyError(code)
    if cause is not None:
        error.__cause__ = cause
    return error


def _metadata(path: Path, prefix: str) -> os.stat_result | None:
    try:
        value = path.lstat()
    except FileNotFoundError:
        return None
    except OSError as exc:
        raise _error(f"{prefix}_METADATA_FAILED", exc)
    if stat.S_ISLNK(value.st_mode):
        raise _error(f"{prefix}_SYMLINK")
    return value


def _project_root(project_root: Path) -> Path:
    lexical = Path(os.path.abspath(os.fspath(project_root)))
    metadata = _metadata(lexical, "MIGRATION_PLAN_PROJECT_ROOT")
    if metadata is not None and not stat.S_ISDIR(metadata.st_mode):
        raise _error("MIGRATION_PLAN_PROJECT_ROOT_NOT_DIRECTORY")
    return lexical.resolve(strict=False)


def resolve_database(project_root: Path, database_input: str | Path) -> tuple[Path, str]:
    root = _project_root(project_root)
    raw = str(database_input).strip()
    encoded = raw.encode("utf-8", errors="strict")
    if not raw or "\x00" in raw or len(encoded) > MAXIMUM_PATH_BYTES:
        raise _error("MIGRATION_PLAN_DATABASE_PATH_INVALID")
    candidate = Path(raw)
    joined = candidate if candidate.is_absolute() else root / candidate
    lexical = Path(os.path.abspath(os.fspath(joined)))
    try:
        relative = lexical.relative_to(root)
    except ValueError as exc:
        raise _error("MIGRATION_PLAN_DATABASE_PATH_ESCAPE", exc)
    if relative == Path(".") or not relative.parts:
        raise _error("MIGRATION_PLAN_DATABASE_PATH_INVALID")

    current = root
    for part in relative.parts[:-1]:
        current /= part
        metadata = _metadata(current, "MIGRATION_PLAN_DATABASE_PARENT")
        if metadata is None:
            break
        if not stat.S_ISDIR(metadata.st_mode):
            raise _error("MIGRATION_PLAN_DATABASE_PARENT_NOT_DIRECTORY")
    _metadata(lexical, "MIGRATION_PLAN_DATABASE")
    selected = lexical.resolve(strict=False)
    try:
        selected.relative_to(root)
    except ValueError as exc:
        raise _error("MIGRATION_PLAN_DATABASE_PATH_ESCAPE", exc)
    logical = relative.as_posix()
    return selected, logical


def canonical_request_bytes(project_root: Path, database_input: str | Path) -> bytes:
    _database, logical = resolve_database(project_root, database_input)
    return json.dumps(
        {"route": ROUTE, "database": logical},
        ensure_ascii=False,
        sort_keys=True,
        separators=(",", ":"),
    ).encode("utf-8")


def rust_argv(project_root: Path, database_input: str | Path) -> tuple[str, ...]:
    _database, logical = resolve_database(project_root, database_input)
    return ("migration", "plan", str(_project_root(project_root)), logical.encode("utf-8").hex())


def request_digest(project_root: Path, database_input: str | Path) -> str:
    return hashlib.sha256(canonical_request_bytes(project_root, database_input)).hexdigest()


def _reject_sidecars(database: Path) -> None:
    for suffix in _SIDECAR_SUFFIXES:
        sidecar = Path(f"{database}{suffix}")
        metadata = _metadata(sidecar, "MIGRATION_PLAN_SIDECAR")
        if metadata is not None:
            if not stat.S_ISREG(metadata.st_mode):
                raise _error("MIGRATION_PLAN_SIDECAR_NOT_FILE")
            raise _error("MIGRATION_PLAN_SIDECAR_PRESENT")


def _identity(database: Path) -> tuple[int, int, int, int]:
    metadata = _metadata(database, "MIGRATION_PLAN_DATABASE")
    if metadata is None:
        raise _error("MIGRATION_PLAN_DATABASE_DISAPPEARED")
    if not stat.S_ISREG(metadata.st_mode):
        raise _error("MIGRATION_PLAN_DATABASE_NOT_FILE")
    return (
        int(metadata.st_dev),
        int(metadata.st_ino),
        int(metadata.st_size),
        int(metadata.st_mtime_ns),
    )


def _open(database: Path) -> sqlite3.Connection:
    identity = _identity(database)
    if identity[2] > MAXIMUM_DATABASE_BYTES:
        raise _error("MIGRATION_PLAN_DATABASE_TOO_LARGE")
    connection: sqlite3.Connection | None = None
    try:
        connection = sqlite3.connect(
            f"{database.as_uri()}?mode=ro&immutable=1",
            uri=True,
            timeout=0.0,
            isolation_level=None,
        )
        connection.row_factory = sqlite3.Row
        connection.execute("PRAGMA query_only=ON")
        connection.execute("PRAGMA trusted_schema=OFF")
        query_only = connection.execute("PRAGMA query_only").fetchone()
    except sqlite3.Error as exc:
        if connection is not None:
            connection.close()
        raise _error("MIGRATION_PLAN_DATABASE_OPEN_FAILED", exc)
    if query_only is None or int(query_only[0]) != 1:
        connection.close()
        raise _error("MIGRATION_PLAN_QUERY_ONLY_FAILED")
    return connection


def _table_exists(connection: sqlite3.Connection) -> bool:
    try:
        row = connection.execute(
            "SELECT 1 FROM sqlite_master WHERE type='table' AND name='syntavra_schema_migrations'"
        ).fetchone()
    except sqlite3.Error as exc:
        raise _error("MIGRATION_PLAN_SCHEMA_QUERY_FAILED", exc)
    return row is not None


def _migration_rows(connection: sqlite3.Connection) -> list[dict[str, Any]]:
    if not _table_exists(connection):
        return []
    try:
        columns = tuple(
            (str(row[1]), str(row[2]).upper(), int(row[3]), int(row[5]))
            for row in connection.execute("PRAGMA table_info(syntavra_schema_migrations)")
        )
        if columns != _EXPECTED_COLUMNS:
            raise _error("MIGRATION_PLAN_SCHEMA_COLUMNS_INVALID")
        rows = [
            {
                "version": int(row[0]),
                "name": str(row[1]),
                "identity": str(row[2]),
                "applied_at": float(row[3]),
            }
            for row in connection.execute(
                "SELECT version,name,identity,applied_at "
                "FROM syntavra_schema_migrations ORDER BY version"
            )
        ]
    except MigrationPlanReadOnlyError:
        raise
    except (sqlite3.Error, TypeError, ValueError, OverflowError) as exc:
        raise _error("MIGRATION_PLAN_ROWS_INVALID", exc)

    versions: set[int] = set()
    for row in rows:
        version = row["version"]
        name = row["name"]
        identity = row["identity"]
        applied_at = row["applied_at"]
        if version < 1 or version in versions:
            raise _error("MIGRATION_PLAN_VERSION_INVALID")
        versions.add(version)
        if not name or len(name.encode("utf-8")) > 1024:
            raise _error("MIGRATION_PLAN_NAME_INVALID")
        if _IDENTITY_RE.fullmatch(identity) is None:
            raise _error("MIGRATION_PLAN_IDENTITY_INVALID")
        if not math.isfinite(applied_at) or applied_at < 0:
            raise _error("MIGRATION_PLAN_APPLIED_AT_INVALID")
    return rows


def migration_plan_read_only_result(
    project_root: Path,
    database_input: str | Path,
) -> dict[str, Any]:
    database, logical = resolve_database(project_root, database_input)
    metadata = _metadata(database, "MIGRATION_PLAN_DATABASE")
    if metadata is None:
        _reject_sidecars(database)
        return {
            "database": logical,
            "current_version": 0,
            "target_version": 0,
            "pending": [],
        }
    if not stat.S_ISREG(metadata.st_mode):
        raise _error("MIGRATION_PLAN_DATABASE_NOT_FILE")

    _reject_sidecars(database)
    before = _identity(database)
    connection = _open(database)
    try:
        rows = _migration_rows(connection)
    finally:
        connection.close()
    _reject_sidecars(database)
    if _identity(database) != before:
        raise _error("MIGRATION_PLAN_SOURCE_CHANGED")

    current = max((int(row["version"]) for row in rows), default=0)
    return {
        "database": logical,
        "current_version": current,
        "target_version": current,
        "pending": [],
    }


__all__ = [
    "CAPABILITY",
    "CONTRACT_VERSION",
    "INPUT_FORMAT",
    "INPUT_PROFILE",
    "MAXIMUM_DATABASE_BYTES",
    "MAXIMUM_PATH_BYTES",
    "MigrationPlanReadOnlyError",
    "ROUTE",
    "SCHEMA_VERSION",
    "canonical_request_bytes",
    "migration_plan_read_only_result",
    "request_digest",
    "resolve_database",
    "rust_argv",
]

File: crates/syntavra-cli/src/migration_plan_read_only_contract.rs
use std::collections::BTreeSet;
use std::fs;
use std::path::{Component, Path, PathBuf};
use std::time::UNIX_EPOCH;

use rusqlite::{Connection, OpenFlags};
use serde_json::json;

const MAXIMUM_DATABASE_BYTES: u64 = 64 * 1024 * 1024;
const MAXIMUM_PATH_BYTES: usize = 4096;
const SIDECAR_SUFFIXES: &[&str] = &["-journal", "-shm", "-wal"];
const EXPECTED_COLUMNS: &[(&str, &str, i64, i64)] = &[
    ("version", "INTEGER", 0, 1),
    ("name", "TEXT", 1, 0),
    ("identity", "TEXT", 1, 0),
    ("applied_at", "REAL", 1, 0),
];

#[derive(Debug, Clone, PartialEq, Eq)]
struct FileIdentity {
    bytes: u64,
    modified_nanos: u128,
}

fn metadata(path: &Path, prefix: &str) -> Result<Option<fs::Metadata>, String> {
    match fs::symlink_metadata(path) {
        Ok(value) => {
            if value.file_type().is_symlink() {
                return Err(format!("{prefix}_SYMLINK"));
            }
            Ok(Some(value))
        }
        Err(error) if error.kind() == std::io::ErrorKind::NotFound => Ok(None),
        Err(_) => Err(format!("{prefix}_METADATA_FAILED")),
    }
}

fn validate_logical_path(value: &str) -> Result<PathBuf, String> {
    if value.is_empty() || value.len() > MAXIMUM_PATH_BYTES || value.contains('\0') {
        return Err("MIGRATION_PLAN_DATABASE_PATH_INVALID".to_owned());
    }
    let path = Path::new(value);
    if path.is_absolute() {
        return Err("MIGRATION_PLAN_DATABASE_PATH_INVALID".to_owned());
    }
    let mut normalized = PathBuf::new();
    for component in path.components() {
        match component {
            Component::Normal(part) => normalized.push(part),
            _ => return Err("MIGRATION_PLAN_DATABASE_PATH_INVALID".to_owned()),
        }
    }
    if normalized.as_os_str().is_empty() {
        return Err("MIGRATION_PLAN_DATABASE_PATH_INVALID".to_owned());
    }
    Ok(normalized)
}

fn database_path(project_root: &Path, logical: &str) -> Result<PathBuf, String> {
    if let Some(value) = metadata(project_root, "MIGRATION_PLAN_PROJECT_ROOT")? {
        if !value.is_dir() {
            return Err("MIGRATION_PLAN_PROJECT_ROOT_NOT_DIRECTORY".to_owned());
        }
    }
    let relative = validate_logical_path(logical)?;
    let mut current = project_root.to_path_buf();
    let parts = relative.components().collect::<Vec<_>>();
    for component in parts.iter().take(parts.len().saturating_sub(1)) {
        if let Component::Normal(part) = component {
            current.push(part);
            match metadata(&current, "MIGRATION_PLAN_DATABASE_PARENT")? {
                Some(value) if !value.is_dir() => {
                    return Err("MIGRATION_PLAN_DATABASE_PARENT_NOT_DIRECTORY".to_owned())
                }
                Some(_) => {}
                None => break,
            }
        }
    }
    Ok(project_root.join(relative))
}

fn reject_sidecars(database: &Path) -> Result<(), String> {
    let database_text = database.to_string_lossy();
    for suffix in SIDECAR_SUFFIXES {
        let sidecar = PathBuf::from(format!("{database_text}{suffix}"));
        if let Some(value) = metadata(&sidecar, "MIGRATION_PLAN_SIDECAR")? {
            if !value.is_file() {
                return Err("MIGRATION_PLAN_SIDECAR_NOT_FILE".to_owned());
            }
            return Err("MIGRATION_PLAN_SIDECAR_PRESENT".to_owned());
        }
    }
    Ok(())
}

fn source_identity(database: &Path) -> Result<FileIdentity, String> {
    let value = metadata(database, "MIGRATION_PLAN_DATABASE")?
        .ok_or_else(|| "MIGRATION_PLAN_DATABASE_DISAPPEARED".to_owned())?;
    if !value.is_file() {
        return Err("MIGRATION_PLAN_DATABASE_NOT_FILE".to_owned());
    }
    let modified_nanos = value
        .modified()
        .map_err(|_| "MIGRATION_PLAN_DATABASE_METADATA_FAILED".to_owned())?
        .duration_since(UNIX_EPOCH)
        .map_err(|_| "MIGRATION_PLAN_DATABASE_METADATA_FAILED".to_owned())?
        .as_nanos();
    Ok(FileIdentity {
        bytes: value.len(),
        modified_nanos,
    })
}

fn open_database(database: &Path) -> Result<Connection, String> {
    let identity = source_identity(database)?;
    if identity.bytes > MAXIMUM_DATABASE_BYTES {
        return Err("MIGRATION_PLAN_DATABASE_TOO_LARGE".to_owned());
    }
    let flags = OpenFlags::SQLITE_OPEN_READ_ONLY
        | OpenFlags::SQLITE_OPEN_NO_MUTEX
        | OpenFlags::SQLITE_OPEN_URI;
    let connection = Connection::open_with_flags(database, flags)
        .map_err(|_| "MIGRATION_PLAN_DATABASE_OPEN_FAILED".to_owned())?;
    connection
        .execute_batch("PRAGMA query_only=ON; PRAGMA trusted_schema=OFF;")
        .map_err(|_| "MIGRATION_PLAN_QUERY_ONLY_FAILED".to_owned())?;
    let query_only: i64 = connection
        .query_row("PRAGMA query_only", [], |row| row.get(0))
        .map_err(|_| "MIGRATION_PLAN_QUERY_ONLY_FAILED".to_owned())?;
    if query_only != 1 {
        return Err("MIGRATION_PLAN_QUERY_ONLY_FAILED".to_owned());
    }
    Ok(connection)
}

fn table_exists(connection: &Connection) -> Result<bool, String> {
    let value: i64 = connection
        .query_row(
            "SELECT EXISTS(SELECT 1 FROM sqlite_master WHERE type='table' AND name='syntavra_schema_migrations')",
            [],
            |row| row.get(0),
        )
        .map_err(|_| "MIGRATION_PLAN_SCHEMA_QUERY_FAILED".to_owned())?;
    Ok(value == 1)
}

fn validate_identity(value: &str) -> bool {
    value.len() == 64
        && value
            .bytes()
            .all(|byte| byte.is_ascii_digit() || (b'a'..=b'f').contains(&byte))
}

fn current_version(connection: &Connection) -> Result<i64, String> {
    if !table_exists(connection)? {
        return Ok(0);
    }
    let mut columns = connection
        .prepare("PRAGMA table_info(syntavra_schema_migrations)")
        .map_err(|_| "MIGRATION_PLAN_SCHEMA_QUERY_FAILED".to_owned())?;
    let actual = columns
        .query_map([], |row| {
            Ok((
                row.get::<_, String>(1)?,
                row.get::<_, String>(2)?.to_ascii_uppercase(),
                row.get::<_, i64>(3)?,
                row.get::<_, i64>(5)?,
            ))
        })
        .map_err(|_| "MIGRATION_PLAN_SCHEMA_QUERY_FAILED".to_owned())?
        .collect::<Result<Vec<_>, _>>()
        .map_err(|_| "MIGRATION_PLAN_SCHEMA_QUERY_FAILED".to_owned())?;
    let expected = EXPECTED_COLUMNS
        .iter()
        .map(|(name, kind, not_null, primary_key)| {
            (
                (*name).to_owned(),
                (*kind).to_owned(),
                *not_null,
                *primary_key,
            )
        })
        .collect::<Vec<_>>();
    if actual != expected {
        return Err("MIGRATION_PLAN_SCHEMA_COLUMNS_INVALID".to_owned());
    }

    let mut statement = connection
        .prepare(
            "SELECT version,name,identity,applied_at FROM syntavra_schema_migrations ORDER BY version",
        )
        .map_err(|_| "MIGRATION_PLAN_SCHEMA_QUERY_FAILED".to_owned())?;
    let rows = statement
        .query_map([], |row| {
            Ok((
                row.get::<_, i64>(0)?,
                row.get::<_, String>(1)?,
                row.get::<_, String>(2)?,
                row.get::<_, f64>(3)?,
            ))
        })
        .map_err(|_| "MIGRATION_PLAN_ROWS_INVALID".to_owned())?;
    let mut versions = BTreeSet::new();
    let mut current = 0_i64;
    for row in rows {
        let (version, name, identity, applied_at) =
            row.map_err(|_| "MIGRATION_PLAN_ROWS_INVALID".to_owned())?;
        if version < 1 || !versions.insert(version) {
            return Err("MIGRATION_PLAN_VERSION_INVALID".to_owned());
        }
        if name.is_empty() || name.len() > 1024 {
            return Err("MIGRATION_PLAN_NAME_INVALID".to_owned());
        }
        if !validate_identity(&identity) {
            return Err("MIGRATION_PLAN_IDENTITY_INVALID".to_owned());
        }
        if !applied_at.is_finite() || applied_at < 0.0 {
            return Err("MIGRATION_PLAN_APPLIED_AT_INVALID".to_owned());
        }
        current = version;
    }
    Ok(current)
}

pub fn migration_plan_json(project_root: &str, logical_database: &str) -> Result<String, String> {
    let root = Path::new(project_root);
    let database = database_path(root, logical_database)?;
    reject_sidecars(&database)?;
    if metadata(&database, "MIGRATION_PLAN_DATABASE")?.is_none() {
        return serde_json::to_string(&json!({
            "current_version": 0,
            "database": logical_database,
            "pending": [],
            "target_version": 0,
        }))
        .map_err(|_| "MIGRATION_PLAN_JSON_FAILED".to_owned());
    }

    let before = source_identity(&database)?;
    let connection = open_database(&database)?;
    let current = current_version(&connection)?;
    drop(connection);
    reject_sidecars(&database)?;
    if source_identity(&database)? != before {
        return Err("MIGRATION_PLAN_SOURCE_CHANGED".to_owned());
    }
    serde_json::to_string(&json!({
        "current_version": current,
        "database": logical_database,
        "pending": [],
        "target_version": current,
    }))
    .map_err(|_| "MIGRATION_PLAN_JSON_FAILED".to_owned())
}
